### PR TITLE
JEP 413: External Snippets and Snippet Attributes Support

### PR DESCRIPTION
**Pull request by mishrasandeep**
https://issues.apache.org/jira/browse/NETBEANS-6493. Added support for external Snippets, Snippet Attributes - region, lang, class, file, id and Copy Snippets functionality.

### CODE REVIEW

File: ide/editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java [hunk 1]
```diff
@@ -451,8 +458,26 @@ private class MouseEventListener extends MouseAdapter {
...
+                    getTextToCopy(HtmlDoc, text, snippetElement);   
+                    handleTextToCopy(HtmlDoc, snippetElement, text);                        
+                    systemClipboard.setContents(new StringSelection(text.toString()), null);
+                    view.getHighlighter().removeAllHighlights(); 
+                    try {
+                        view.getHighlighter().addHighlight(e.getSourceElement().getStartOffset(), e.getSourceElement().getEndOffset(), new DefaultHighlightPainter(Color.GRAY));
+                    } catch (BadLocationException ex) {
+                        Exceptions.printStackTrace(ex);
+                    }
+                }else if (e != null && HyperlinkEvent.EventType.ACTIVATED.equals(e.getEventType())) {
```
Comment (lahodaj) [inline ide/editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java:480]: Nit: formatting.

File: ide/editor.completion/src/org/netbeans/modules/editor/completion/HTMLDocView.java [hunk 2]
```diff
@@ -123,7 +126,7 @@ public void run(){
                     Document doc = getDocument();
                     doc.remove(0, doc.getLength());
                     getEditorKit().read(in, getDocument(), 0);  //!!! still too expensive to be called from AWT
-                    setCaretPosition(0);
+                    setCaretPosition(0);
```
Comment (lahodaj) [inline ide/editor.completion/src/org/netbeans/modules/editor/completion/HTMLDocView.java:129]: Nit: noop/unnecessary change.

File: java/java.editor/test/unit/data/org/netbeans/modules/java/editor/javadocsnippet/data/HighlightTag.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ * A simple program.
+ * {@link System#out}
+ * {@snippet :
+ * class HelloWorld {
+ *     public static void main(String... args) {
+ *         System.out.println("Hello World!");      // @highlight substring="println"
+ *     }
+ * }
+ * }
+ */    private static void method(R r) {
```
Comment (lahodaj) [inline java/java.editor/test/unit/data/org/netbeans/modules/java/editor/javadocsnippet/data/HighlightTag.java:14]: Nit: formatting.

File: java/java.source.base/src/org/netbeans/modules/java/source/save/CasualDiff.java [hunk 4]
```diff
@@ -4740,6 +4741,10 @@ private int diffDocTree(DCDocComment doc, DCTree oldT, DCTree newT, int[] elemen
 //                  }
                     break;
                 }
+//                if(oldT.getKind().toString().equals(TreeShims.SNIPPET)){
```
Comment (lahodaj) [inline java/java.source.base/src/org/netbeans/modules/java/source/save/CasualDiff.java:4744]: Nit: cleanup unnecessary comments.

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/MarkupTag.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package org.netbeans.api.java.source.ui.snippet;
```
Comment (lahodaj) [inline java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/MarkupTag.java:1]: License header?

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/MarkupTagAttribute.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package org.netbeans.api.java.source.ui.snippet;
```
Comment (lahodaj) [inline java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/MarkupTagAttribute.java:1]: License header?

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/SnippetMarkupTagExtractor.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
+package org.netbeans.api.java.source.ui.snippet;
```
Comment (lahodaj) [inline java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/SnippetMarkupTagExtractor.java:1]: License header?

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/SnippetTagCommentParser.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
+package org.netbeans.api.java.source.ui.snippet;
```
Comment (lahodaj) [inline java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/SnippetTagCommentParser.java:1]: License header?

File: java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/SourceLineMeta.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package org.netbeans.api.java.source.ui.snippet;
```
Comment (lahodaj) [inline java/java.sourceui/src/org/netbeans/api/java/source/ui/snippet/SourceLineMeta.java:1]: License header?

**Review by lahodaj — APPROVED** [review]
Overall, looks OK. But please go through the patch, fix formatting, remove unnecessary changes, add license headers as needed and please make sure RAT passes - it does not pass currently:
https://app.travis-ci.com/github/apache/netbeans/jobs/556893196

**Comment by mishrasandeep** [comment]
Tests for Link Tag will be updated with PR https://github.com/apache/netbeans/pull/3884

**Comment by mishrasandeep** [comment]
Tests for Link Tag will be updated with PR https://github.com/apache/netbeans/pull/3884
